### DISCUSSION TITLE
Does the track method of Reaction have unnecessary batch operations?

### DISCUSSION THREAD

**Discussion by dSalieri** [mobxjs/mobx#3754] [Q&A]
I watched code base of `Reaction` class and some his dependencies and I've seen the next:

```JavaScript
/// I omitted detailed code
track() {
  startBatch();
  trackDerivedFunction(this, fn, undefined);
  endBatch();
}

function trackDerivedFunction(derivation, f, context) {
  globalState.inBatch++;
  f.call(context);
  globalState.inBatch--;
}
```

So my question is the next: Why does the `track` method have batch operations, if we have operations `inBatch++`,  `inBatch--` inside `trackDerivedFunction`?

If `startBatch` and `endBatch` has important sense in `track`, please explain it, desirable with examples on real examples.

**Comment by mweststrate** [comment]
Sorry, most of it has been written 5+ years ago, so the details are not top
of mind. (Why care?) probably has to do with updates not immediately
starting new reactions but waiting until the current one is finished. If
you want to know more, I recommend to disable the lines and run the unit
tests, likely some will fail.

**Reply by dSalieri** [reply]
@mweststrate  Only one operation runs the `track` method (via `onInvalidate_`) and this is `runReaction_` (it also has `startBatch` and `endBatch`). So I think the `track` method has useless batch operations (my opinion) and probably `trackDerivedFunction` also has unnesessary `inBatch` increment and decrement.

**Tell me how I can run the tests you are talking about.**

P.S I am investigating mobx architecture and I what to know how this works.

**Comment by mweststrate** [comment]
Check out the repo, yarn, yarn test

**Reply by dSalieri** [reply]
Hmm, interesting... At first glance it seems that there are unnecessary batchings, but in fact everything is to the point.

It's sad that there aren't comments for some parts of code base of mobx

P.S Thanks that you responded me.
